### PR TITLE
Add full and incremental backup information to `list-backup`

### PR DESCRIPTION
**Pull request by bleidens**
Add a new label to each backup in the `list-backup` output that references the backup type considering:
 - R: for rsync backup
 - F: for full pg_basebackup
 - I: for incremental pg_basebackup

If a backup doesn't have a parent backup nor the backup method set to rsync, it's labeled as a full backup. This also adds a new field "backup_type" to the JSON output, containing the type label.

To get the correct backup type without relying on volatile information, a new property `backup_type` was added to the `LocalBackupInfo` object, which returns the label for `backup.info` at backup time instead of relying on the configuration at runtime. 

References: BAR-207

### CODE REVIEW

**Comment by barthisrael** [comment]
Rebased this on top of `master`.

**Comment by barthisrael** [comment]
The implementation seems to work fine.

I'm reposting the conversation that we have in our internal chat here just for historical purposes.

The changes from this PR seems like this when the backup has a custom name:

```console
$ barman list-backup all
pg17 20240725T183731 - F - 'full_2' - Thu Jul 25 16:37:33 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB
pg17 20240725T183723 - I - 'incremental_2' - Thu Jul 25 16:37:24 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183718 - I - 'incremental_1' - Thu Jul 25 16:37:19 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183702 - F - 'full' - Thu Jul 25 16:37:05 2024 - Size: 251.3 MiB - WAL Size: 32.0 MiB
support 20240725T183751 - R - 'rsync_incremental' - Thu Jul 25 16:37:55 2024 - Size: 235.2 MiB - WAL Size: 0 B
support 20240725T183740 - R - 'rsync_full' - Thu Jul 25 16:37:47 2024 - Size: 235.2 MiB - WAL Size: 32.0 Mi
```

I think it should show like this instead:

```console
$ barman list-backup all
pg17 20240725T183731 'full_2' - 'F' - Thu Jul 25 16:37:33 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB
pg17 20240725T183723 'incremental_2' - 'I' - Thu Jul 25 16:37:24 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183718 'incremental_1' - 'I' - Thu Jul 25 16:37:19 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183702 'full' - 'F' - Thu Jul 25 16:37:05 2024 - Size: 251.3 MiB - WAL Size: 32.0 MiB
support 20240725T183751 'rsync_incremental' - 'R' - Thu Jul 25 16:37:55 2024 - Size: 235.2 MiB - WAL Size: 0 B
support 20240725T183740 'rsync_full' - 'R' - Thu Jul 25 16:37:47 2024 - Size: 235.2 MiB - WAL Size: 32.0 MiB
```

Even though that seems a bit broken visually, this remains consistent with the output previous to the the changes introduced by the PR, which would look like this:

```console
$ barman list-backup all
pg17 20240725T183731 'full_2' - Thu Jul 25 16:37:33 2024 - Size: 251.3 MiB - WAL Size: 48.0 MiB
pg17 20240725T183723 'incremental_2' - Thu Jul 25 16:37:24 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183718 'incremental_1' - Thu Jul 25 16:37:19 2024 - Size: 35.3 MiB - WAL Size: 32.0 MiB
pg17 20240725T183702 'full' - Thu Jul 25 16:37:05 2024 - Size: 251.3 MiB - WAL Size: 32.0 MiB
support 20240725T183751 'rsync_incremental' - Thu Jul 25 16:37:55 2024 - Size: 235.2 MiB - WAL Size: 0 B
support 20240725T183740 'rsync_full' - Thu Jul 25 16:37:47 2024 - Size: 235.2 MiB - WAL Size: 32.0 MiB
```

File: barman/output.py [hunk 1]
```diff
@@ -676,11 +676,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
...
+        else:
+            backup_type = "F"
+
+        out_list = [
+            "%s %s - %s - "
+            % (backup_info.server_name, backup_info.backup_id, backup_type)
+        ]
         if backup_info.backup_name is not None:
             out_list.append("'%s' - " % backup_info.backup_name)
-        else:
```
Comment (martinmarques) [inline barman/output.py:682]: Why was this removed?

File: barman/output.py [hunk 2]
```diff
@@ -676,11 +676,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
             self.info(backup_info.backup_id)
             return
 
-        out_list = ["%s %s " % (backup_info.server_name, backup_info.backup_id)]
+        # Backup type can be either full, incremental or rsync
+        if backup_info.backup_manager.config.backup_method == "rsync":
```
Comment (barthisrael) [inline barman/output.py:680]: We cannot use the config object from the manager because it contains the current value of `backup_method` for the server.

For example, if you do the following you will have a buggy output:

* Configure server with method `rsync`
* Take a backup
* Change configuration to use method `postgres`
* Execute `barman list-backup all`

The output will show the label `F`, when in fact the backup was taken with `rsync`.

In other words, do not rely on configuration to set the label, but on the info stored in the `backup.info` at backup time.

**Review by barthisrael — CHANGES_REQUESTED** [review]

File: barman/output.py [hunk 3]
```diff
@@ -676,11 +676,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
...
 
-        out_list = ["%s %s " % (backup_info.server_name, backup_info.backup_id)]
+        out_list = ["%s %s" % (backup_info.server_name, backup_info.backup_id)]
+
         if backup_info.backup_name is not None:
-            out_list.append("'%s' - " % backup_info.backup_name)
+            out_list.append(" '%s'" % backup_info.backup_name)
+
+        # Backup type can be either full, incremental or rsync
+        if backup_info.config.backup_method == "rsync":
```
Comment (barthisrael) [inline barman/output.py:685]: I can see you changed this, but this is still volatile as it keeps depending on the configuration.

See this as an example -- `support` server was originally configured with `rsync`:

```console
$ grep -i method /etc/barman.d/support.conf
#backup_method = rsync
backup_method = postgres
```

The `barman list-backup` command will wrongly show them as "full":

```console
$ barman list-backup support
support 20240725T183751 'rsync_incremental' - F - Thu Jul 25 16:37:55 2024 - Size: 235.2 MiB - WAL Size: 0 B
support 20240725T183740 'rsync_full' - F - Thu Jul 25 16:37:47 2024 - Size: 235.2 MiB - WAL Size: 32.0 MiB
```

File: barman/output.py [hunk 4]
```diff
@@ -1470,13 +1479,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
...
-            backup_id=backup_info.backup_id,
-        )
+        output = dict(backup_id=backup_info.backup_id)
 
         if backup_info.backup_name is not None:
             output.update({"backup_name": backup_info.backup_name})
 
+        # Set backup type label
+        if backup_info.config.backup_method == "rsync":
+            backup_type = "RSYNC"
```
Comment (barthisrael) [inline barman/output.py:1489]:
```suggestion
            backup_type = "rsync"
```

File: barman/output.py [hunk 5]
```diff
@@ -1470,13 +1479,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
...
+        elif backup_info.is_incremental:
+            backup_type = "INCREMENTAL"
```
Comment (barthisrael) [inline barman/output.py:1491]:
```suggestion
            backup_type = "incremental"
```

File: barman/output.py [hunk 6]
```diff
@@ -1470,13 +1479,20 @@ def result_list_backup(self, backup_info, backup_size, wal_size, retention_statu
...
+        else:
+            backup_type = "FULL"
```
Comment (barthisrael) [inline barman/output.py:1493]:
```suggestion
            backup_type = "full"
```

**Review by barthisrael — CHANGES_REQUESTED** [review]
The issue with the `list-backup` output remains as it still depends on volatile information. That logic should depend on the information taken from the `backup.info` file rather than from the configuration file.

Besides that, what do you think of using lower case values for the backup type in JSON output? I thought of that after taking a look at `barman diagnose` output, where almost everything is lower case. I've put suggestions to change the code in that sense.

Last but not least, I was wondering if we should have some function in `LocalBackupInfo` class to tell us the type of the backup -- something like this:

```python
    def get_backup_type(self, short):
```

When `short` is `True`, returns either `R`, `F` or `I`.
When `short` is `False`, returns either `rsync`, `full` or `incremental`.

That way we could avoid "code duplication" (we already have some as we perform the same logic in both output writers), and that could even be reused later. For example: what if we decide to add that information to `barman diagnose` later as well? The method could be reused instead of copying and pasting the logic a third time.

**Comment by barthisrael** [comment]
About my previous comment, maybe `get_backup_type` could have no parameters, or even be a property `backup_type`, which would always return either `full`, `rsync` or `incremental.

I guess the only place where we will use the "short" version is in `list-backup`. So, there we could simply convert text to upper case and select only the first character.

File: barman/infofile.py [hunk 7]
```diff
@@ -751,6 +751,22 @@ def has_children(self):
...
 
+    @property
+    def backup_type(self):
+        """
+        Returns a string with the backup type label.
+
+        The backup type can be one of the following:
+        - "rsync": If the backup mode is "rsync".
+        - "incremental": If the mode is "postgres" and the backup is incremental.
+        - "full": If the mode is "postgres" and the backup is not incremental.
```
Comment (barthisrael) [inline barman/infofile.py:762]:
```suggestion
        - ``rsync``: If the backup mode is "rsync``.
        - ``incremental``: If the mode is ``postgres`` and the backup is incremental.
        - ``full``: If the mode is ``postgres`` and the backup is not incremental.
```

Comment (barthisrael) [reply barman/infofile.py:762]: Sphinx recommendation for literals.

File: barman/output.py (hunk 1 above)
Comment (barthisrael) [reply barman/output.py:682]: We talked about this in the internal chat.
It was removed because of the way the output string is built.

File: tests/test_output.py [hunk 8]
```diff
@@ -1176,7 +1174,38 @@ def test_result_list_backup_with_backup_name(self, capsys):
...
+        console_writer.init_list_backup(backup_info.server_name, False)
+        console_writer.result_list_backup(backup_info, 12345, 54321, "test status")
+        console_writer.close()
+
+        # Capture and assert output
+        out, _err = capsys.readouterr()
+        assert (
+            f"{backup_info.server_name} {backup_info.backup_id} - {expected_backup_type} -"
+            in out
+        )
```
Comment (barthisrael) [inline tests/test_output.py:1208]: I'd say to remove these unit tests from here, and actually write unit tests for `LocalBackupInfo.backup_type` property in `test_infofile.py`.

What I mean is: in `barman.output` we simply use the returning value of the aforementioned property.

If you want to check that `barman -f console list-backup` uses `LocalBackupInfo.backup_type` as expected, that's surely a valid test. But you need to property and check what `list-backup` does.

For example:

```python
@pytest.mark.parametrize(
    ("backup_type", "expected_value"),
    (
        ("some_type", "S"),
        ("another_type", "A"),
    )
)
def test_result_list_backup_types(...):
    ...
    # Call the following method with a mock on `backup_type` property
    writer.result_list_backup(...)
    # assert that result_list_backup played with "backup_type" and printed "expected_value"

File: tests/test_output.py [hunk 9]
```diff
@@ -1857,6 +1887,33 @@ def test_result_list_backup_with_backup_name(self, capsys):
...
+            writer = output.JsonOutputWriter()
+            writer.init_list_backup(backup_info.server_name, False)
+            writer.result_list_backup(backup_info, 12345, 54321, "test status")
+            writer.close()
+            (out, err) = capsys.readouterr()
+            json_output = json.loads(out)
+            backup = find_by_attr(
+                json_output[backup_info.server_name], "backup_id", backup_info.backup_id
+            )
+            assert backup["backup_type"] == expected_type
```
Comment (barthisrael) [inline tests/test_output.py:1915]: Similar comment here regarding the unit test.

**Review by barthisrael — CHANGES_REQUESTED** [review]
I have mocked some of my local `backup.info` files to see different outputs from `barman list-backup` command. It looks good!

* With console output writer:

```console
$ barman list-backup all
WARNING: No backup strategy set for server 'pg17' (using default 'concurrent_backup').
pg17 20240725T183731 'full_2' - F - Thu Jul 25 16:37:33 2024 - Size: 251.3 MiB - WAL Size: 80.0 MiB - WAITING_FOR_WALS
pg17 20240725T183723 'incremental_2' - I - FAILED
pg17 20240725T183718 - I - Thu Jul 25 16:37:19 2024 - Size: 35.3 MiB - WAL Size: 64.0 MiB
pg17 20240725T183702 'full' - F - Thu Jul 25 16:37:05 2024 - Size: 251.3 MiB - WAL Size: 32.0 MiB
support 20240725T183751 'rsync_incremental' - R - Thu Jul 25 16:37:55 2024 - Size: 235.2 MiB - WAL Size: 0 B
support 20240725T183740 'rsync_full' - R - Thu Jul 25 16:37:47 2024 - Size: 235.2 MiB - WAL Size: 32.0 MiB
```

* With JSON output writer:

```console
$ barman -f json list-backup all | jq
{
  "_WARNING": [
    "No backup strategy set for server 'pg17' (using default 'concurrent_backup')."
  ],
  "pg17": [
    {
      "backup_id": "20240725T183731",
      "backup_name": "full_2",
      "backup_type": "full",
      "end_time": "Thu Jul 25 16:37:33 2024",
      "end_time_timestamp": "1721925453",
      "retention_status": "-",
      "size": "251.3 MiB",
      "size_bytes": 263536050,
      "status": "WAITING_FOR_WALS",
      "wal_size": "80.0 MiB",
      "wal_size_bytes": 83886080
    },
    {
      "backup_id": "20240725T183723",
      "backup_name": "incremental_2",
      "backup_type": "incremental",
      "status": "FAILED"
    },
    {
      "backup_id": "20240725T183718",
      "backup_type": "incremental",
      "end_time": "Thu Jul 25 16:37:19 2024",
      "end_time_timestamp": "1721925439",
      "retention_status": "-",
      "size": "35.3 MiB",
      "size_bytes": 37065043,
      "status": "DONE",
      "wal_size": "64.0 MiB",
      "wal_size_bytes": 67108864
    },
    {
      "backup_id": "20240725T183702",
      "backup_name": "full",
      "backup_type": "full",
      "end_time": "Thu Jul 25 16:37:05 2024",
      "end_time_timestamp": "1721925425",
      "retention_status": "-",
      "size": "251.3 MiB",
      "size_bytes": 263534269,
      "status": "DONE",
      "wal_size": "32.0 MiB",
      "wal_size_bytes": 33554432
    }
  ],
  "support": [
    {
      "backup_id": "20240725T183751",
      "backup_name": "rsync_incremental",
      "backup_type": "rsync",
      "end_time": "Thu Jul 25 16:37:55 2024",
      "end_time_timestamp": "1721925475",
      "retention_status": "-",
      "size": "235.2 MiB",
      "size_bytes": 246574350,
      "status": "DONE",
      "wal_size": "0 B",
      "wal_size_bytes": 0
    },
    {
      "backup_id": "20240725T183740",
      "backup_name": "rsync_full",
      "backup_type": "rsync",
      "end_time": "Thu Jul 25 16:37:47 2024",
      "end_time_timestamp": "1721925467",
      "retention_status": "-",
      "size": "235.2 MiB",
      "size_bytes": 246574343,
      "status": "DONE",
      "wal_size": "32.0 MiB",
      "wal_size_bytes": 33554432
    }
  ]
}
```

The code looks good!
I posted a few suggestions regarding unit tests.

We also need to add unit tests for the `backup_property` in `test_infofile.py`.

File: tests/test_output.py [hunk 10]
```diff
@@ -1814,6 +1850,7 @@ def test_result_list_backup(self, capsys):
 
         backup = find_by_attr(json_output[bi.server_name], "backup_id", bi.backup_id)
         assert bi.backup_id == backup["backup_id"]
+        assert backup["backup_type"] == "rsync"
```
Comment (barthisrael) [inline tests/test_output.py:1853]:
```suggestion
```

I think we can remove this as this part is tested in the new unit test (`test_result_list_backup_types`)

**Review by barthisrael — CHANGES_REQUESTED** [review]

**Review by barthisrael — APPROVED** [review]
Looks good!
